src/lib/reasoners/adt_rel.ml

```diff
@@ -622,25 +639,58 @@ let pick_best ds =
   in
   Some (r, c)
 
+let pick_enum ds uf =
+  Domains.fold_enums
+    (fun r best ->
+       let rr, _ = Uf.find_r uf r in
+       let d = Domains.get r ds in
+       let cd = Domain.cardinal d in
+       match Th.embed rr, best with
+       | Constr _, _ -> best
+       | _, Some (n, _, _) when n <= cd -> best
+       | _ ->
+         let c = Domain.choose d in
+         Some (cd, r, c)
+    ) ds None
+
+let split_enum env uf =
+  let ds = Uf.(GlobalDomains.find (module Domains) @@ domains uf) in
+  let* n, r, c = pick_enum ds uf in
+  let n = Numbers.Q.from_int n in
+  if can_split env n then
+    let _, cons = Option.get @@ build_constr_eq r c in
+    let nr, ctx = X.make cons in
+    (* In the current implementation of `X.make`, we produce
+       a nonempty context only for interpreted semantic values
+       of the `Arith` and `Records` theories. The semantic
+       values `cons` never involves such values. *)
+    assert (Lists.is_empty ctx);
+    Some (LR.mkv_eq r nr)
+  else
+    None
```

Comment:
I think there is some duplication with `split_best_domain` that can be eliminated by defining a `make_split n r c` function that makes the split, and a `pick` function that dispatches to either `pick_enum` or `pick_best` as appropriate.